jmdaemon/jmdaemon/onionmc.py

```diff
@@ -0,0 +1,1214 @@
+                # since dns will never reach this point without an active
+                # connection.
+                reactor.callLater(0.0, temp_p.try_to_connect)
+            return temp_p
+        else:
+            p = self.get_peer_by_id(temp_p.peer_location())
+            if overwrite_connection:
+                if connection:
+                    log.info("Updating status to connected for peer {}.".format(temp_p.peer_location()))
+                    p.update_status(PEER_STATUS_CONNECTED)
+                else:
+                    p.update_status(PEER_STATUS_DISCONNECTED)
+            if with_nick:
+                p.set_nick(nick)
+            return p
+
+    def get_all_connected_peers(self) -> list:
+        return self.get_connected_directory_peers() + \
+               self.get_connected_nondirectory_peers()
+
+    def get_connected_directory_peers(self) -> list:
+        return [p for p in self.peers if p.directory and p.status() == \
+                PEER_STATUS_HANDSHAKED]
+
+    def get_connected_nondirectory_peers(self) -> list:
+        return [p for p in self.peers if (not p.directory) and p.status() == \
+                PEER_STATUS_HANDSHAKED]
+
+    def wait_for_directories(self) -> None:
+        # Notice this is checking for *handshaked* dps;
+        # the handshake will have been initiated once a
+        # connection was seen:
+        if len(self.get_connected_directory_peers()) == 0:
+            return
+        # This is what triggers the start of taker/maker workflows.
+        if not self.on_welcome_sent:
+            self.on_welcome(self)
+            self.on_welcome_sent = True
+            self.wait_for_directories_loop.stop()
+
+    """ CONTROL MESSAGES SENT BY US
+    """
+    def send_peers(self, requesting_peer: OnionPeer,
+                   peer_filter: List[OnionPeer]) -> None:
+        """ This message is sent by directory peers, currently
+        only when a privmsg has to be forwarded to them. It
+        could also be sent by directories to non-directory peers
+        according to some other algorithm.
+        If peer_filter is specified, only those peers will be sent.
+        The peerlist message should have this format:
+        (1) entries comma separated
+        (2) each entry is serialized nick then the NICK_PEERLOCATOR_SEPARATOR
+            then host:port
+        (3) Peers that do not have a reachable location are not sent.
+        """
+        if not requesting_peer.status() == PEER_STATUS_HANDSHAKED:
+            raise OnionPeerConnectionError(
+                "Cannot send peer list to unhandshaked peer")
+        peerlist = set()
+        for p in self.get_connected_nondirectory_peers():
+            # don't send a peer to itself
+            if p == requesting_peer:
+                continue
+            if len(peer_filter) > 0 and p not in peer_filter:
+                continue
+            if not p.status() == PEER_STATUS_HANDSHAKED:

```

Comment:
Nit in 5448a2631163c9bba5ca722a0628b33af717cdcd:

```suggestion
            if p.status() != PEER_STATUS_HANDSHAKED:
```

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/5ba9e0bc9731aad29bbdf282c8fa5b027b7556cd